File: Sources/Accord.Audio/Tools.cs
// Accord Audio Library
// The Accord.NET Framework
// http://accord-framework.net
//
// Copyright © César Souza, 2009-2017
// cesarsouza at gmail.com
//
//    This library is free software; you can redistribute it and/or
//    modify it under the terms of the GNU Lesser General Public
//    License as published by the Free Software Foundation; either
//    version 2.1 of the License, or (at your option) any later version.
//
//    This library is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
//    Lesser General Public License for more details.
//
//    You should have received a copy of the GNU Lesser General Public
//    License along with this library; if not, write to the Free Software
//    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
//

namespace Accord.Audio
{
    using Accord.Math;
    using System;
    using System.Runtime.InteropServices;
    using System.Collections.Generic;
    using Accord.Compat;
    using System.Numerics;

    /// <summary>
    ///   Tool functions for audio processing.
    /// </summary>
    /// 
    public static class Tools
    {
        /// <summary>
        ///   Interleaves the channels into a single array.
        /// </summary>
        /// 
        public static float[] Interleave(this float[][] channels)
        {
            if (channels == null) throw new ArgumentNullException("channels");
            if (channels.Length == 0) return new float[0];

            int c = channels.Length;
            int n = channels[0].Length;

            float[] data = new float[c * n];

            for (int i = 0, k = 0; i < n; i++)
                for (int j = 0; j < c; j++)
                    data[k++] = channels[j][i];

            return data;
        }

        /// <summary>
        ///   Interleaves the channels into a single array.
        /// </summary>
        /// 
        public static float[] Interleave(this float[,] channels)
        {
            float[] result = new float[channels.Length];
            Buffer.BlockCopy(channels, 0, result, 0, result.Length * sizeof(float));
            return result;
        }

        /// <summary>
        ///   Obsolete. Please use <see cref="Accord.Math.Transforms.FourierTransform2.GetMagnitudeSpectrum(Complex[])"/> instead.
        /// </summary>
        /// 
        [Obsolete("Please use Accord.Math.Transforms.FourierTransform2.GetMagnitudeSpectrum(fft) instead.")]
        public static double[] GetMagnitudeSpectrum(Complex[] fft)
        {
            return Accord.Math.Transforms.FourierTransform2.GetMagnitudeSpectrum(fft);
        }

        /// <summary>
        ///   Obsolete. Please use <see cref="Accord.Math.Transforms.FourierTransform2.GetPowerSpectrum(Complex[])"/> instead.
        /// </summary>
        /// 
        [Obsolete("Please use Accord.Math.Transforms.FourierTransform2.GetPowerSpectrum(fft) instead.")]
        public static double[] GetPowerSpectrum(Complex[] fft)
        {
            return Accord.Math.Transforms.FourierTransform2.GetPowerSpectrum(fft);
        }

        /// <summary>
        ///   Obsolete. Please use <see cref="Accord.Math.Transforms.FourierTransform2.GetPhaseSpectrum(Complex[])"/> instead.
        /// </summary>
        /// 
        [Obsolete("Please use Accord.Math.Transforms.FourierTransform2.GetPhaseSpectrum(fft) instead.")]
        public static double[] GetPhaseSpectrum(Complex[] fft)
        {
            return Accord.Math.Transforms.FourierTransform2.GetPhaseSpectrum(fft);
        }

        /// <summary>
        ///   Obsolete. Please use <see cref="Accord.Math.Transforms.FourierTransform2.GetFrequencyVector(int, int)"/> instead.
        /// </summary>
        /// 
        [Obsolete("Please use Accord.Math.Transforms.FourierTransform2.GetFrequencyVector(length, sampleRate) instead.")]
        public static double[] GetFrequencyVector(int length, int sampleRate)
        {
            return Accord.Math.Transforms.FourierTransform2.GetFrequencyVector(length, sampleRate);

        }

        /// <summary>
        ///   Obsolete. Please use <see cref="Accord.Math.Transforms.FourierTransform2.GetSpectralResolution(int, int)"/> instead.
        /// </summary>
        /// 
        [Obsolete("Please use Accord.Math.Transforms.FourierTransform2.GetSpectralResolution(samplingRate, samples) instead.")]
        public static double GetSpectralResolution(int samplingRate, int samples)
        {
            return Accord.Math.Transforms.FourierTransform2.GetSpectralResolution(samplingRate, samples);
        }

        /// <summary>
        ///   Obsolete. Please use <see cref="Accord.Math.Transforms.FourierTransform2.GetPowerCepstrum(Complex[])"/> instead.
        /// </summary>
        /// 
        [Obsolete("Please use Accord.Math.Transforms.FourierTransform2.GetPowerCepstrum(signal) instead.")]
        public static double[] GetPowerCepstrum(Complex[] signal)
        {
            return Accord.Math.Transforms.FourierTransform2.GetPowerCepstrum(signal);
        }

        /// <summary>
        ///   Computes the Root-Mean-Square (RMS) value of the given samples.
        /// </summary>
        /// 
        /// <param name="samples">The samples.</param>
        /// 
        /// <returns>The root-mean-square value of the samples.</returns>
        /// 
        public static double RootMeanSquare(this float[] samples)
        {
            return RootMeanSquare(samples, 0, samples.Length);
        }

        /// <summary>
        ///   Computes the Root-Mean-Square (RMS) value of the given samples.
        /// </summary>
        /// 
        /// <param name="samples">The samples.</param>
        /// <param name="startIndex">The start index.</param>
        /// <param name="count">The number of samples, starting at start index, to compute.</param>
        /// 
        /// <returns>The root-mean-square value of the samples.</returns>
        /// 
        public static double RootMeanSquare(this float[] samples, int startIndex, int count)
        {
            float sum = 0;
            for (int i = 0; i < count; i++)
            {
                float s = samples[startIndex + i];
                sum += s * s;
            }

            return Math.Sqrt(sum);
        }

        /// <summary>
        ///   Computes the maximum value of the given samples.
        /// </summary>
        /// 
        /// <param name="samples">The samples.</param>
        /// 
        /// <returns>The maximum value of the samples</returns>
        /// 
        public static float Max(this float[] samples)
        {
            return Max(samples, 0, samples.Length);
        }

        /// <summary>
        ///   Computes the maximum value of the given samples.
        /// </summary>
        /// 
        /// <param name="samples">The samples.</param>
        /// <param name="startIndex">The start index.</param>
        /// <param name="count">The number of samples, starting at start index, to compute.</param>
        /// 
        /// <returns>The maximum value of the samples</returns>
        /// 
        public static float Max(this float[] samples, int startIndex, int count)
        {
            float max = 0;
            for (int i = 0; i < count; i++)
            {
                if (samples[i + startIndex] > max)
                    max = samples[i];
            }

            return max;
        }

        /// <summary>
        ///   Finds the peaks of a signal.
        /// </summary>
        /// 
        /// <param name="samples">The samples.</param>
        /// 
        /// <returns>The index of the peaks found in the sample.</returns>
        /// 
        public static int[] FindPeaks(this double[] samples)
        {
            var peaks = new List<int>();

            for (int i = 1; i < samples.Length - 1; i++)
            {
                if (samples[i] > samples[i - 1] && samples[i] > samples[i + 1])
                    peaks.Add(i);
            }

            return peaks.ToArray();
        }

        /// <summary>
        ///   Finds the peaks of a signal.
        /// </summary>
        /// 
        /// <param name="samples">The samples.</param>
        /// 
        /// <returns>The index of the peaks found in the sample.</returns>
        /// 
        public static int[] FindPeaks(this float[] samples)
        {
            var peaks = new List<int>();

            for (int i = 1; i < samples.Length - 1; i++)
            {
                if (samples[i] > samples[i - 1] && samples[i] > samples[i + 1])
                    peaks.Add(i);
            }

            return peaks.ToArray();
        }

        /// <summary>
        ///   Obsolete. Please use the extension method from the Accord namespace instead.
        /// </summary>
        /// 
        [Obsolete("Please use the extension method from the Accord namespace instead.")]
        public static byte[] ToByteArray<T>(T value)
            where T : struct
        {
            return ExtensionMethods.ToByteArray(value);
        }

        /// <summary>
        ///   Obsolete. Please use <see cref="ExtensionMethods.ToStruct{T}(byte[], int)"/>
        /// </summary>
        /// 
        [Obsolete("Please use ToStruct<T>() instead.")]
        public static T RawDeserialize<T>(this byte[] rawData)
            where T : struct
        {
            return rawData.ToStruct<T>();
        }

        /// <summary>
        ///   Obsolete. Please use <see cref="ExtensionMethods.ToStruct{T}(byte[], int)"/>
        /// </summary>
        /// 
        [Obsolete("Please use ToStruct<T>() instead.")]
        public static T RawDeserialize<T>(this byte[] rawData, int position)
            where T : struct
        {
            return rawData.ToStruct<T>(position);
        }
    }
}

File: Sources/Extras/Accord.Video.FFMPEG.GPL/Tools.cpp
// Accord FFMPEG Library
// The Accord.NET Framework
// http://accord-framework.net
//
// Copyright © César Souza, 2009-2017
// cesarsouza at gmail.com
//
//    This program is free software; you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation; either version 2 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program; if not, write to the Free Software
//    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include "StdAfx.h"
#include "Tools.h"

#include <string>

using namespace System;
using namespace System::IO;
using namespace Accord::Audio;
using namespace Accord::Video;
using namespace Accord::Compat;

namespace Accord {
    namespace Video {
        namespace FFMPEG
        {
            int64_t frameToPTS(AVStream* stream, int64_t frame)
            {
                return (int64_t(frame) * stream->r_frame_rate.den * stream->time_base.den)
                    / (int64_t(stream->r_frame_rate.num) * stream->time_base.num);
            }

            int64_t ptsToFrame(AVStream* stream, int64_t pts)
            {
                return (int64_t(pts) * stream->time_base.num *  stream->codec->time_base.num) /
                    (int64_t(stream->time_base.den) * stream->codec->time_base.den);
            }

            /* check that a given sample format is supported by the encoder */
            static int check_sample_fmt(AVCodec* codec, enum ::AVSampleFormat& sample_fmt)
            {
                const enum ::AVSampleFormat* p = codec->sample_fmts;
                while (*p != AV_SAMPLE_FMT_NONE)
                {
                    if (*p == sample_fmt)
                        return 1;
                    p++;
                }
                return 0;
            }

            /* check that a given sample rate is supported by the encoder */
            static int check_sample_rate(AVCodec *codec, int sampleRate)
            {
                const int* p = codec->supported_samplerates;

                if (p == 0)
                    return 1;

                while (*p) {
                    if (*p == sampleRate)
                        return 1;
                    p++;
                }
                return 0;
            }

            String^ str(const char* str)
            {
                return gcnew System::String(str);
            }

            String^ str(int64_t ts, AVRational* tb)
            {
                if (ts == AV_NOPTS_VALUE)
                    return "NOPTS";
                return String::Format("{0}", av_q2d(*tb) * ts);
            }

            void CHECK(int errorCode, String^ message)
            {
                if (errorCode < 0)
                {
                    // TODO: Use av_err2str
                    char buff[AV_ERROR_MAX_STRING_SIZE];
                    av_make_error_string(&buff[0], AV_ERROR_MAX_STRING_SIZE, errorCode);
                    System::String^ ffmpegErrorMessage = System::Runtime::InteropServices::Marshal::PtrToStringAnsi((IntPtr)&buff[0]);
                    throw gcnew VideoException(String::Format("{0}. Error {1}: {2}. See console output for more details.",
                        message, errorCode, ffmpegErrorMessage));
                }
            }

            const char* str2native(String^ str, char* native)
            {
                if (str == nullptr)
                    return nullptr;

                IntPtr ptr = System::Runtime::InteropServices::Marshal::StringToHGlobalUni(str);
                wchar_t* nativeFileNameUnicode = (wchar_t*)ptr.ToPointer();
                int utf8StringSize = WideCharToMultiByte(CP_UTF8, 0, nativeFileNameUnicode, -1, nullptr, 0, nullptr, nullptr);
                WideCharToMultiByte(CP_UTF8, 0, nativeFileNameUnicode, -1, native, utf8StringSize, nullptr, nullptr);
                return native;
            }


            int64_t TimeSpanToPTS(TimeSpan& duration, AVStream* stream, AVCodecContext* codec)
            {
                uint64_t hundredNanoseconds = duration.Ticks;
                AVRational tickRate = { 1, 10000000 };
                AVRational streamRatio = stream->time_base;
                AVRational codecRatio = codec->time_base;
                int64_t streamPos = av_rescale_q_rnd(hundredNanoseconds, tickRate, streamRatio, AVRounding::AV_ROUND_UP);
                int64_t result = av_rescale_q_rnd(streamPos, streamRatio, codecRatio, AVRounding::AV_ROUND_UP);
                return result;
            }

            TimeSpan PTSToTimeSpan(uint64_t pts, AVStream* stream, AVCodecContext* codec)
            {
                if (pts == AV_NOPTS_VALUE)
                    pts = 0;

                AVRational codecRatio = codec->time_base;
                AVRational streamRatio = stream->time_base;
                AVRational tickRate = { 1, 10000000 };

                int64_t codecPos = av_rescale_q_rnd(pts, codecRatio, streamRatio, AVRounding::AV_ROUND_DOWN);
                uint64_t  hundredNanoseconds = av_rescale_q_rnd(codecPos, streamRatio, tickRate, AVRounding::AV_ROUND_DOWN);

                return TimeSpan::FromTicks(hundredNanoseconds);
            }


            enum ::AVSampleFormat s2f(const Accord::Audio::SampleFormat& format)
            {
                switch (format)
                {
                case Accord::Audio::SampleFormat::Format16Bit:
                    return AV_SAMPLE_FMT_S16;
                case Accord::Audio::SampleFormat::Format32Bit:
                    return AV_SAMPLE_FMT_S32;
                case Accord::Audio::SampleFormat::Format32BitIeeeFloat:
                    return AV_SAMPLE_FMT_FLT;
                }

                throw gcnew ArgumentOutOfRangeException("format", "Invalid audio format.");
            }

            enum ::AVPixelFormat p2f(const System::Drawing::Imaging::PixelFormat& format)
            {
                switch (format)
                {
                case System::Drawing::Imaging::PixelFormat::Format8bppIndexed:
                    return AV_PIX_FMT_GRAY8;
                case System::Drawing::Imaging::PixelFormat::Format24bppRgb:
                    return AV_PIX_FMT_BGR24;
                case System::Drawing::Imaging::PixelFormat::Format32bppArgb:
                    return AV_PIX_FMT_BGRA;
                }

                throw gcnew ArgumentOutOfRangeException("format", "Invalid image format.");
            }


            void check_redistributable()
            {
                static bool checked = false;

                if (checked)
                    return;
#if NET35
                bool is64bits = Accord::Compat::EnvironmentEx::Is64BitProcess;
                DirectoryInfo^ windowPath = Directory::GetParent(Environment::GetFolderPath(Environment::SpecialFolder::System));

                if (is64bits)
                {
                    throw gcnew InvalidOperationException("This application cannot be run in 64-bits.");
                }
                else
                {
                    bool success = windowPath->GetDirectories("WinSxS\\x86_Microsoft.VC90.CRT*")->Length > 0;

                    if (!success)
                    {
                        throw gcnew InvalidOperationException("This application requires the x86 version of " +
                            "the Microsoft Visual C++ 2008 SP1 Redistributable Package to be installed on this computer. Please " +
                            "download and install it from https://www.microsoft.com/en-us/download/details.aspx?id=5582");
                    }
                }
#else
                String^ system32Path = Environment::GetFolderPath(Environment::SpecialFolder::SystemX86);
                String^ system64Path = Environment::GetFolderPath(Environment::SpecialFolder::System);
                bool is64bits = Environment::Is64BitProcess;

                String^ msvcDllName = "msvcp140.dll";

                bool success = File::Exists(Path::Combine(system32Path, msvcDllName));

                if (is64bits)
                {
                    success &= File::Exists(Path::Combine(system64Path, msvcDllName));
                }

                if (!success)
                {
                    if (is64bits)
                    {
                        throw gcnew InvalidOperationException("This application requires both the x64 and x86 versions of " +
                            "the Visual C++ Redistributable for Visual Studio 2015 to be installed on this computer. Please " +
                            "download and install them from https://www.microsoft.com/en-us/download/details.aspx?id=48145");
                    }
                    else
                    {
                        throw gcnew InvalidOperationException("This application requires the x86 version of " +
                            "the Visual C++ Redistributable for Visual Studio 2015 to be installed on this computer. Please " +
                            "download and install it from https://www.microsoft.com/en-us/download/details.aspx?id=48145");
                    }
                }
#endif

                checked = true;
            }

        }


    }
}
